lib/tensorflow/session.rb

```diff
@@ -23,96 +23,26 @@ def initialize()
   # * *Returns* :
   #   - nil
   #
   def run(inputs, outputs, targets)
-  	inputNames = Tensorflow::String_Vector.new
-  	inputValues = Tensorflow::Tensor_Vector.new
-    if inputs != nil
-  	  inputs.each do |key, value|
-  		  inputValues.push(value)
-  		  inputNames.push(key)
-  	  end
-    end
-
-  	outputNames = Tensorflow::String_Vector.new
-  	outputs.each do |name|
-  		outputNames.push(name)
-  	end
+    input_names, input_values = initialize_inputs(inputs)
+    output_names, output_values = initialize_outputs(outputs)
+    target_names = initialize_targets(targets)
 
-  	targetNames = Tensorflow::String_Vector.new
-    if targets != nil
-  	  targets.each do |name|
-  		  targetNames.push(name)
-  	  end
-    end
-
-  	outputValues = Tensorflow::Tensor_Vector.new
    	status = Tensorflow::TF_NewStatus()
-	  Tensorflow::TF_Run_wrapper(self.session , inputNames, inputValues, outputNames, outputValues, targetNames, self.status)
+	  Tensorflow::TF_Run_wrapper(self.session, input_names, input_values, output_names, output_values, target_names, self.status)
     raise ("Incorrect specifications passed.")  if Tensorflow::TF_GetCode(status) != Tensorflow::TF_OK
+
     output_array = []
-    (0..outputValues.size - 1).each do |i|
-      size = Tensorflow::tensor_size(outputValues[i])
-      type = Tensorflow::TF_TensorType(outputValues[i])
-
-      case type
-      when Tensorflow::TF_DOUBLE
-        c_array = Tensorflow::Double.new(size)
-        Tensorflow::double_reader(outputValues[i], c_array, size)
-      when Tensorflow::TF_INT64
-        c_array = Tensorflow::Long_long.new(size)
-        Tensorflow::long_long_reader(outputValues[i], c_array, size)
-      when Tensorflow::TF_INT32
-        c_array = Tensorflow::Int.new(size)
-        Tensorflow::int_reader(outputValues[i], c_array, size)
-      when Tensorflow::TF_COMPLEX128
-        c_array = Tensorflow::complex_reader(outputValues[i])
-      else
-        raise "Data type not supported."
-      end
 
-      num_dimensions = Tensorflow::TF_NumDims(outputValues[i])
-      out_dimension = []
-      (0..num_dimensions - 1).each do |j|
-         out_dimension.push(Tensorflow::TF_Dim(outputValues[i], j))
-      end
-      output = []
-      (0..size - 1).each do |j|
-         output.push(c_array[j])
-       end
-       out_dimension.reverse!
-       (0..out_dimension.length - 2).each do |k|
-          total_dim = output.length
-          dim_array = []
-          temp_array = []
-          (0..total_dim - 1).each do |ind|
-            temp_array.push(output[ind])
-            dim_array.push(temp_array) if (temp_array.length == out_dimension[k])
-            temp_array = []            if (temp_array.length == out_dimension[k])
-          end
-          output = dim_array
-       end
-       output_array.push(output)
+    output_values.each do |value|
+      converted_value = convert_value_for_output_array(value)
+      output_array.push(converted_value)
     end
+
     output_array
```

Comment:
I tried this, but was surprised to find that some objects being used that implement `each` in this class don't actually implement `map`. I'm still learning about ruby interfacing with C so didn't want to look into it too much, but since `each` is already implemented, I think it would probably be a quick fix to get enumerable methods, either by including `Enumerable` directly or creating a wrapper class.